src/ring_theory/tensor_product.lean

```diff
@@ -1,43 +1,292 @@
+Authors: Scott Morrison, Johan Commelin
 -/
 
 import linear_algebra.tensor_product
-import algebra.algebra.basic
+import algebra.algebra.tower
 
 /-!
 # The tensor product of R-algebras
 
-We construct the R-algebra structure on `A ⊗[R] B`, when `A` and `B` are both `R`-algebras,
-and provide the structure isomorphisms
+Let `R` be a (semi)ring and `A` an `R`-algebra.
+In this file we:
 
-* `R ⊗[R] A ≃ₐ[R] A`
-* `A ⊗[R] R ≃ₐ[R] A`
-* `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+- Define the `A`-module structure on `A ⊗ M`, for an `R`-module `M`.
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+
+## Main declaration
+
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
 -/
 
 universes u v₁ v₂ v₃ v₄
 
-namespace algebra
-
 open_locale tensor_product
 open tensor_product
 
 namespace tensor_product
 
+variables {R A M N P : Type*}
+
+/-!
+### The `A`-module structure on `A ⊗[R] M`
+-/
+
+open linear_map
+open algebra (lsmul)
+
+section defn
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [module R M] [module A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [module R N]
+
+variables (R A M N)
+
+@[priority 500]
+instance algebra_tensor_module : module A (M ⊗[R] N) :=
+{ smul := λ a, (lsmul R M a).rtensor N,
+  one_smul := λ x, by simp only [alg_hom.map_one, one_eq_id, id.def, id_coe, rtensor_id],
+  mul_smul := λ a b x, by simp only [alg_hom.map_mul, rtensor_mul, mul_apply],
+  smul_add := λ a x y, by simp only [map_add],
+  smul_zero := λ a, by simp only [map_zero],
+  add_smul := λ a b x, by simp only [alg_hom.map_add, add_apply, rtensor_add],
+  zero_smul := λ x, by simp only [rtensor_zero, alg_hom.map_zero, zero_apply] }
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.algebra_tensor_module R R M N =
+          @tensor_product.module R _ M N _ _ _ _ := rfl
+
+end defn
+
+namespace algebra_tensor_module
+
+section semiring
+variables [comm_semiring R] [semiring A] [algebra R A]
+variables [add_comm_monoid M] [module R M] [module A M] [is_scalar_tower R A M]
+variables [add_comm_monoid N] [module R N]
+variables [add_comm_monoid P] [module A P]
+
+variables {R A M}
+
+lemma smul_def (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+@[simp] lemma smul_tmul (c : A) (x : M) (y : N) : c • (x ⊗ₜ[R] y) = (c • x) ⊗ₜ[R] y := rfl
+
+@[ext] theorem ext {g h : (M ⊗[R] N) →ₗ[A] P}
+  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
+linear_map.ext $ λ z, tensor_product.induction_on z (by simp_rw linear_map.map_zero) H $
+λ x y ihx ihy, by rw [g.map_add, h.map_add, ihx, ihy]
```

Comment:
This would be a better ext lemma if you could state it as:
```
@[ext]
theorem curry'_inj {g h : M ⊗[R] N →ₗ[A] P}
  (H : curry' g = curry' h) : g = h :=
by {
  have := λ x, linear_map.congr_fun (linear_map.congr_fun H x),
  sorry -- your proof here
}
```

Reply:
I think there's an easy proof of this using `linear_map.restrict_scalars_injective`; can you merge master so that that's available?

Reply:
Is this definitely a better ext lemma? I'll yield the point immediately if you still prefer it, as I'd love to get this merged.

Also, I just realised `curry'_inj` is already included in these changes: https://github.com/leanprover-community/mathlib/pull/4773/files#diff-40a1b7b0f1a7171e81d65a33e7c47bb428d521828f394e4bb5b4b9b70e82c657R142

In any case I've added https://github.com/leanprover-community/mathlib/pull/4773/commits/b97f3424a353241fa77147de84f808ec969f9170 with two lemmas I used in my proof:
```lean
begin
  suffices : curry (g.restrict_scalars R) = curry (h.restrict_scalars R),
  { exact linear_map.restrict_scalars_injective R (curry_inj this), },
  simp only [← restrict_scalars_curry', H],
end
```
before I realised `curry'_inj` already exists.

Reply:
Yes, it's a better ext lemma because it handles a tower of towers correctly. For the same reason, `tensor_product.ext` is not actually marked `@[ext]`!

Reply:
The proof of `curry_injective` I was looking for is:
```
@[ext]
lemma curry'_injective : function.injective (curry' : (M ⊗ N →ₗ[A] P) → (M →ₗ[A] N →ₗ[R] P)) :=
λ x y h, linear_map.restrict_scalars_injective R $ curry_inj $
  (congr_arg (linear_map.restrict_scalars R) h : _)
```

Reply:
Which gives you a shorter proof for `ext`:
```
theorem ext {g h : (M ⊗[R] N) →ₗ[A] P}
  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
curry'_inj $ linear_map.ext₂ $ H
```

Reply:
Thanks, I'm actually working on this right now. Some fixes required after moving the `ext` but should be done shortly.